File: 44wildcardMatching/wildcardMatching.py
class Solution:
    def isMatch(self, s: str, p: str) -> bool:
        dp = [True, *itertools.repeat(False, len(p))]
        for column in range(1, len(dp)): dp[column] = dp[column - 1] and p[column - 1] == '*'        
        for row in range(1, len(s) + 1):
            prev = [*dp]
            dp[0] = False
            for column in range(1, len(dp)):
                if s[row - 1] == p[column - 1] or p[column - 1] == '?': dp[column] = prev[column - 1]
                elif p[column - 1] == '*': dp[column] = dp[column - 1] or prev[column]
                else: dp[column] = False
        return dp[-1]

File: 44wildcardMatching/wildcardMatching.java
class Solution {
    public boolean isMatch(String s, String p) {
        boolean[] dp = new boolean[p.codePointCount(0, p.length()) + 1];
        dp[0] = true;
        IntStream.range(1, dp.length).forEach(column -> dp[column] = dp[column - 1] && p.codePointAt(column - 1) == "*".codePointAt(0));
        IntStream.range(1, s.codePointCount(0, s.length()) + 1).forEach(row -> {
            final boolean[] prev = dp.clone();
            dp[0] = false;
            IntStream.range(1, dp.length).forEach(column -> {
                if (s.codePointAt(row - 1) == p.codePointAt(column - 1) || p.codePointAt(column - 1) == "?".codePointAt(0)) dp[column] = prev[column - 1];
                else if (p.codePointAt(column - 1) == "*".codePointAt(0)) dp[column] = dp[column - 1] || prev[column];
                else dp[column] = false;});});
        return dp[dp.length - 1];
    }
}
